packages/apify/src/actor.ts

```diff
@@ -268,13 +270,42 @@ export class Actor {
             customAfterSleepMillis = this.config.get('metamorphAfterSleepMillis'),
             ...metamorphOpts
         } = options;
+
         const runId = this.config.get('actorRunId')!;
+
         await this.apifyClient.run(runId).metamorph(targetActorId, input, metamorphOpts);
 
         // Wait some time for container to be stopped.
         await sleep(customAfterSleepMillis);
     }
 
+    /**
+     * Internally reboots this actor. The system stops the current container and starts
+     * a new container with the same run ID.
+     *
+     * @param [_options]
+     * @ignore
+     */
+    async reboot(_options: RebootOptions = {}): Promise<void> {
+        if (!this.isAtHome()) {
+            log.warning('Actor.reboot() is only supported when running on the Apify platform.');
+            return;
+        }
+
+        // Waiting for all the listeners to finish, as `.metamorph()` kills the container.
+        await Promise.all(
+            [
+                // `persistState` for individual RequestLists, RequestQueue... instances to be persisted
+                ...events.listeners(ACTOR_EVENT_NAMES_EX.PERSIST_STATE).map((x) => x()),
+                // `migrating` to pause Apify crawlers
+                ...events.listeners(ACTOR_EVENT_NAMES_EX.MIGRATING).map((x) => x()),
```

Comment:
@mnmkng do we need to emit both events here? i believe internally migrating event is just a flavor of persist_state, so the migrating event could be enough? 

https://github.com/apify/apify-ts/blob/master/packages/core/src/events.ts#L104

i was also expecting just `events.emit()` call, the metamorph call waits for 5 minutes before it resolves, which i believe was the reason - to wait for the events to finish. i am not against explicit await like this, but i guess it would be better done in the `metamorph` method directly, and there might still be some other reason for that long sleep inside

Reply:
IIRC `ACTOR_EVENT_NAMES.MIGRATING` is caught only in [`BasicCrawler`](https://github.com/apify/apify-ts/blob/c3850ecdb8024cd9e7e80dbcde97a47a39121968/packages/basic-crawler/src/internals/basic-crawler.ts#L521). 
Don't know what the best practices are, but can imagine a user using only the `RequestList` etc. separately within their own solution - `RequestList` is listening only to the `PERSIST_STATE` event.

Reply:
@B4nan Dunno. I would double check against your new event implementation :)

The long sleep is there so that the container does not abruptly exit while the API performs the metamoph.

Reply:
I guess we need both, as they are fired manually.

Reply:
Also, the sleep call happens only after the `metamorph` API call, when the metamorph is already happening.

I'm just guessing here, but if metamorph call means that the Apify platform performs `docker stop` on our container -> an IPC signal to the main process (`SIGTERM`, grace period, then `SIGKILL`). `SIGTERM`, if uncaught, terminates the node process immediately without waiting for any `async` functions. This is still an uneducated guess, but this sounds like a potential data loss/corruption in case of a long "saving" process and overly eager platform.